[BEAM-8376] Initial version of firestore connector JavaSDK

djelekar opened the pull request:
Added initial version of firestore connector for JavaSDK.

R: @chamikaramj 
------------------------
Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreWriterFn.java:
```diff
@@ -0,0 +1,203 @@
+                }
+
+                for (WriteResult result : future.get()) {
+                    LOG.info("Update time : " + result.getUpdateTime());
```
Probably will be useful to log the firestoreKey here as well.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreWriterFn.java:
```diff
@@ -0,0 +1,203 @@
+                // Break if the commit threw no exception.
+                break;
+            } catch (FirestoreException exception) {
+                if (exception.getCode() == 4) {
```
So 4 means DEADLINE_EXCEEDED ? Please clarify in the comment.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreWriterFn.java:
```diff
@@ -0,0 +1,203 @@
+                // Only log the code and message for potentially-transient errors. The entire exception
+                // will be propagated upon the last retry.
+                LOG.error(
+                        "Error writing batch of {} mutations to Firestore ({}): {}",
```
Log the firestoreKey ?

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/MovingAverage.java:
```diff
@@ -0,0 +1,61 @@
+import org.apache.beam.sdk.transforms.Sum;
+import org.apache.beam.sdk.util.MovingFunction;
+
+class MovingAverage {
```
Please add Javadocs to clarify why we need this.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/AdaptiveThrottlerTest.java:
```diff
@@ -0,0 +1,116 @@
+ * Tests for {@link AdaptiveThrottler}.
+ */
+@RunWith(JUnit4.class)
+public class AdaptiveThrottlerTest {
```
Please also add unit test for the sink. See here for some of the example unit tests that we could add for Firestore.
https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1Test.java

In addition please also consider adding an integration test similar to here: https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/datastore/V1WriteIT.java

djelekar commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreIO.java:
```diff
@@ -0,0 +1,74 @@
+        }
+
+        /**
+         * Returns a new {@link Write} that writes to the Cloud Firestore for the specified collection key.
```
Collection ID is the unique identifier of Firestore collection (e.g. "userSessions"), while Key ID is the unique identifier for a specific document in a collection (e.g. "5a0062ef-3fe2-490d-b1be-06f5bf94a39d").

I added a link to the Firestore [data model page](https://cloud.google.com/firestore/docs/data-model) which explains it pretty well.

Do you think that is enough, or should we adjust naming/docs additionally?

djelekar commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreWriterFn.java:
```diff
@@ -0,0 +1,203 @@
+    /**
+     * Writes a batch of mutations to Cloud Firestore.
+     *
+     * <p>If a commit fails, it will be retried up to {@link #MAX_RETRIES} times. All mutations in
```
This doc is actually false. I've used Firestore "batched writes" which makes writing mutations an atomic operation. So all either succeeds or fails -> no risk of duplicating entries. I updated the docs.

fredzqm commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreBatchRequest.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public WriteBatch batchWithKey(List<T> input, String collection, String key) {
+        WriteBatch batch = firestoreClient.batch();
```
Please note that `WriteBatch` is atomic transaction.
Large `WriteBatch` could leads to contention and high error rates.

We are working on launching a batchWrite API for non-atomic data ingestion use cases. Before it gets launched, the next best option is writing each document separately.

(BTW, we have a max 500 write per commit limit.)

fredzqm replied:
Please make sure this is addressed.

fredzqm commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreBatchRequester.java:
```diff
@@ -0,0 +1,58 @@
+            batch.set(docRef, object);
+        }
+
+        ApiFuture<List<WriteResult>> commit = batch.commit();
```
Please a

fredzqm commented:
Please note that WriteBatch is atomic transaction.
Large WriteBatch could lead to contention and high error rates.

We are working on launching a batchWrite API for non-atomic data ingestion use cases. Before it gets launched, the next best option is writing each document separately.

(BTW, we have a max 500 write per commit limit.)

chamikaramj commented:
Thanks @fredzqm.

@djelekar please let us know when current comments are addressed.

chamikaramj commented:
Any updates ?

Thanks.

djelekar commented:
Hi @chamikaramj , yes the tests are there and I addressed all your comments in code. However, I'm having issues running integration tests on my machine (windows). Once, that is working I'll bump you for one more review.

Hi @fredzqm, I understand your concerns, however, we have better experience running jobs with WriteBatch command. Namely, by doing data imports on large amounts of data, we witnessed faster processing times with WriteBatch then single writes. Can you point me to some documentation or else that aligns with your point?

Additionally, the code utilizes AdaptiveThrottler, with the max batch value of 500, so hopefully, that obeys with the limit.

clement commented:
Hi @djelekar, I work on the Firestore backend, and chiming in to second @fredzqm point. There are two interlocking issues when using atomic WriteBatch for large ingestion (throughput) jobs.

First, under load and based on size, Firestore will split your dataset across multiple servers. When writing atomically to multiple documents, this increase the chance that the write will need to coordinate a 2-phase commit across multiple servers, which will increase the latency of the operation.

Second, Firestore uses a pessimistic locking model under the hood. If the WriteBatch takes longer to execute (because of the issue above, or just because it is doing more work) it will be holding locks longer and disrupt unrelated read/write traffic on the document or index entries.

I can see reasons why the experience looks better with WriteBatch, for example:
- when using single writes, those should be asynchronous, and can (and should) be parallelized more aggressively
- if the ingestion key range is not split, or not actively accessed by other processes, there will initially be no contention and good performance with WriteBatch, however there is a limit to how much throughput you will get from them once the ingestion runs longer and ramps up to more parallelism.

Does that make sense? We are hoping to launch a dedicated feature for writing batches in a non-atomic fashion, but it is unclear at this point when this will be generally available, and as @fredzqm point out, single writes are the best option for now.

willbattel commented:
"any comment"

willbattel commented:
Hey @fredzqm @clement are there any updates regarding the mentioned non-atomic batch writing capability?

aaltay commented:
What is the state of this PR? Do you need help?

willbattel commented:
I noticed the aforementioned bulk writer feature for Firestore has been made available in some clients, such as [Java](https://github.com/googleapis/java-firestore/pull/323) and [Node](https://github.com/googleapis/nodejs-firestore/pull/1252). Can the PR now be continued using this new API for parallelized writes?

oeclint commented:
Does the new batch writer mean we could reopen this?
